### ISSUE TITLE
Test failure with gcc 10.x in Linux

### ISSUE THREAD

**Issue by rhertzog** [cpputest/cpputest#1309]
If you build cpputest with gcc 10.x on Debian Linux on amd64, you will get two test failures:

```
tests/MemoryOperatorOverloadTest.cpp:349: error: Failure in TEST(OutOfMemoryTestsForOperatorNew,
FailingNewArrayOperatorThrowsAnExceptionWhenUsingStdCppNewWithoutOverride)
        expected to throw std::bad_alloc
but threw nothing

.
tests/MemoryOperatorOverloadTest.cpp:344: error: Failure in TEST(OutOfMemoryTestsForOperatorNew,
FailingNewOperatorThrowsAnExceptionWhenUsingStdCppNewWithoutOverride)
        expected to throw std::bad_alloc
but threw nothing
```

Given that GCC 10.x will soon be the default, it would be nice to have fixes for those issues.

The full build log can be seen here: https://people.debian.org/~doko/logs/gcc10-20200225/cpputest_3.8-7_unstable_gcc10.log

This has been reported in Debian first: https://bugs.debian.org/957104

**Comment by rhertzog** [comment]
Also it has been 4 years since the last release and there are several hundreds of commits in master that are newer than the last release.... maybe it would be time to tag a new release ?

**Comment by offa** [comment]
I can confirm this with the current master (GCC 10.0.1 running in a Ubuntu 18.04 docker container) :

```
 Start 23: CppUTestTests.OutOfMemoryTestsForOperatorNew
23/70 Test #23: CppUTestTests.OutOfMemoryTestsForOperatorNew .................................................***Failed    0.00 sec
..
/mnt/tests/CppUTest/MemoryOperatorOverloadTest.cpp:385: error: Failure in TEST(OutOfMemoryTestsForOperatorNew, FailingNewArrayOperatorThrowsAnExceptionWhenUsingStdCppNewWithoutOverride)
	expected to throw std::bad_alloc
but threw nothing
.
/mnt/tests/CppUTest/MemoryOperatorOverloadTest.cpp:380: error: Failure in TEST(OutOfMemoryTestsForOperatorNew, FailingNewOperatorThrowsAnExceptionWhenUsingStdCppNewWithoutOverride)
	expected to throw std::bad_alloc
but threw nothing
...
Errors (2 failures, 762 tests, 6 ran, 6 checks, 0 ignored, 756 filtered out, 0 ms)
```

**Comment by basvodde** [comment]
@rhertzog Yeah, been working on releasing something. Hope to do that in 2 weeks, first been fixing mostly obscure bugs. I'll look at this failure today. I'm surprised by it.

**Comment by basvodde** [comment]
Ok, just ran the tests on Ubuntu Focal, which has gcc-10 and g++-10. Compiled and ran with gcc-10 as default and all tests pass.

**Comment by basvodde** [comment]
Ok. Sorry for not reading clearly. This is not happening on trunk, just for release 3.8 (now replicated). As it is fixed on trunk, this issue basically means "Make a new release", right?

**Comment by rhertzog** [comment]
Ah, sorry, I didn't test the git version and you're right that this report is based on the last released version available in Debian. So I guess, yes, a new release would be nice ;-)

**Comment by basvodde** [comment]
Do you mind if I close this issue then? I'll go over the issues to see which must be fixed for release. I believe you had another open one.

**Comment by rhertzog** [comment]
On Mon, 20 Apr 2020, Bas Vodde wrote:
> Do you mind if I close this issue then? I'll go over the issues to see
> which must be fixed for release. I believe you had another open one.

Actually, I just tried and I get the failures with the master branch just
like @offa said.

I have g++ version "10-20200418-1" (available in Debian Unstable in the
g++-10 package).

$ g++ --version
g++ (Debian 10-20200418-1) 10.0.1 20200418 (experimental) [master revision 27c171775ab:4c277008be0:c5bac7d127f288fd2f8a1f15c3f30da5903141c6]

So it's a newer upstream snapshot dating back to two days ago...

Ubuntu Focal has 10-20200411-0ubuntu1... that makes only a few days of
difference.

But maybe it's not only the compiler version involved here?

Cheers,
-- 
  ⢀⣴⠾⠻⢶⣦⠀   Raphaël Hertzog <hertzog@debian.org>
  ⣾⠁⢠⠒⠀⣿⡁
  ⢿⡄⠘⠷⠚⠋    The Debian Handbook: https://debian-handbook.info/get/
  ⠈⠳⣄⠀⠀⠀⠀   Debian Long Term Support: https://deb.li/LTS

**Comment by basvodde** [comment]
Hmm, so far, I have Ubuntu Focal. When I checkout -v3.8, I get the failure. When I use trunk, I don't. Let me try again, just to be sure (starting Focal again)

**Comment by rhertzog** [comment]
When you say "trunk", you mean the "master" branch, right?

**Comment by basvodde** [comment]
Yes. Saw the failure. Did git checkout origin/master. Run make clean; make tdd. and ... (still compiling)

**Comment by basvodde** [comment]
Failed! :)

**Comment by basvodde** [comment]
Good, wonder why it passed earlier. Anyways, not I'm on it. As it passes on all other compilers and platforms, I'm suspecting a problem in gcc10.

**Comment by basvodde** [comment]
Ok, this is interesting indeed. It seems to be us against the C++ optimizer. The test that fails had failed before due to some optimization. We put a hack in place to fool the compiler optimization. Now... that doesn't work. I've tried an alternative way and that worked.

For if you are interested, the comment above the text refers to a bug report in the clang compiler. It *seems* it is the same compiler optimization. No new bug report done in gcc, I guess the chance they'll fix it is nearly 0.

Comment:

/*
 * CLang 4.2 and memory allocation.
 *
 * Clang 4.2 has done some optimizations to their memory management that actually causes slightly different behavior than what the C++ Standard defines.
 * Usually this is not a problem... but in this case, it is a problem.
 *
 * More information about the optimization can be found at: http://www.open-std.org/jtc1/sc22/wg21/docs/papers/2012/n3433.html
 * We've done a bug-report to clang to fix some of this non-standard behavior, which is open at: http://llvm.org/bugs/show_bug.cgi?id=15541
 *
 * I very much hope that nobody would actually ever hit this bug/optimization as it is hard to figure out what is going on.
 *
 * The original test simply did "new char". Because the memory wasn't assigned to anything and is local in context, the optimization *doesn't* call
 * the operator new overload. Because it doesn't call the operator new (optimizing away a call to operator new), therefore the method wouldn't throw an exception
 * and therefore this test failed.
 *
 * The first attempt to fix this is to create a local variable and assigned the memory to that. Also this doesn't work as it still detects the allocation is
 * local and optimizes away the memory call.
 *
 * Now, we assign the memory on some static global which fools the optimizer to believe that it isn't local and it stops optimizing the operator new call.
 *
 * We (Bas Vodde and Terry Yin) suspect that in a real product, you wouldn't be able to detect the optimization and it's breaking of Standard C++. Therefore,
 * for now, we keep this hack in the test to fool the optimizer and hope nobody will ever notice this 'optimizer behavior' in a real product.
 *
 */

**Comment by basvodde** [comment]
@offa Do you have the Dockerfile of the image where the test fails ?

**Comment by basvodde** [comment]
Ok, submitted a bug report to gcc 10 at: https://gcc.gnu.org/bugzilla/show_bug.cgi?id=94671

**Comment by offa** [comment]
> @offa Do you have the Dockerfile of the image where the test fails ?

Of course, you can pull pre-build images from here: `registry.gitlab.com/offa/docker-images/gcc-10:stable` (it's ready to build cpputest).

The Build steps used:

```sh
mkdir _build && cd _build
cmake -DC++11=ON -DWERROR=ON ..
make
ctest --output-on-failure
```

**Comment by basvodde** [comment]
Ok, fixed with PR #1313 

That was interesting, thanks!

**Comment by offa** [comment]
Seriously, just these few lines and it works!? :laughing:

**Comment by basvodde** [comment]
Actually, this is the long version. The extern declaration is probably not needed but without it, it might lead to compiler warning that a non-declared global variable is being used... 

Just removing "static" would have done.

**Comment by offa** [comment]
Interesting, thanks for the investigation.

**Comment by basvodde** [comment]
It wasn't hard *this* time....  It was the same optimization we debugged years ago and then it took a lot of time. This time, most time was actually spend in properly reproducing it, like adding the docker images. Now we also got the DOS docker image, so we can finally fix the dos builds a bit faster.
